feat(autoware_pointcloud_preprocessor): add diagnostic message

vividf opened the pull request:
## Description
This PR adds diagnostic messages for the main pointcloud preprocessing components, including the Crop Box Filter, Distortion Corrector, and Ring Outlier Filter. The diagnostics include information such as input/output point count, pass rate, processing time, and pipeline latency, enabling real-time monitoring of each module’s status and performance.

## Related links

**Parent Issue:**

- Link

<!-- ⬇️🟢
**Private Links:**

- [CompanyName internal link]()
⬆️🟢 -->

## How was this PR tested?
Sample rosbag. 

### Distortion corrector
![Screenshot from 2025-05-20 12-22-40](https://github.com/user-attachments/assets/3fd1b3ab-f963-40fb-af97-dcf262731a24)


### Concat
![image](https://github.com/user-attachments/assets/829771b6-c358-44df-bf6d-614f21cda817)


### Filter
#### Crop box filter
![Screenshot from 2025-05-19 18-35-44](https://github.com/user-attachments/assets/f91fee9c-c85e-48fb-a056-fcb6a3b64422)


#### Ring outlier filter
![Screenshot from 2025-05-19 16-55-20](https://github.com/user-attachments/assets/40bd35c0-2693-49b2-8cba-931e9049bca5)


 



## Notes for reviewers

None.

## Interface changes

None.

<!-- ⬇️🔴

### Topic changes

#### Additions and removals

| Change type   | Topic Type      | Topic Name    | Message Type        | Description       |
|:--------------|:----------------|:--------------|:--------------------|:------------------|
| Added/Removed | Pub/Sub/Srv/Cli | `/topic_name` | `std_msgs/String`   | Topic description |

#### Modifications

| Version | Topic Type      | Topic Name        | Message Type        | Description       |
|:--------|:----------------|:------------------|:--------------------|:------------------|
| Old     | Pub/Sub/Srv/Cli | `/old_topic_name` | `sensor_msgs/Image` | Topic description |
| New     | Pub/Sub/Srv/Cli | `/new_topic_name` | `sensor_msgs/Image` | Topic description |

### ROS Parameter Changes

#### Additions and removals

| Change type   | Parameter Name | Type     | Default Value | Description       |
|:--------------|:---------------|:---------|:--------------|:------------------|
| Added/Removed | `param_name`   | `double` | `1.0`         | Param description |

#### Modifications

| Version | Parameter Name   | Type     | Default Value | Description       |
|:--------|:-----------------|:---------|:--------------|:------------------|
| Old     | `old_param_name` | `double` | `1.0`         | Param description |
| New     | `new_param_name` | `double` | `1.0`         | Param description |

🔴⬆️ -->

## Effects on system behavior

None.

vividf committed:
feat: add diag msg

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: fix code

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: remove outlier count in ring

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: move format timestamp to utility

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: add paramter to schema

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: add parameter for cluster

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: clean code

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: fix schema

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: move diagnostic updater to filter base class

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: fix schema

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
Merge branch 'main' into feat/add_diag_msg

vividf committed:
Merge branch 'main' into feat/add_diag_msg

vividf committed:
chore: fix spell error

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: set up diagnostic updater

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
Merge branch 'main' into feat/add_diag_msg

vividf committed:
refactor: utilize autoware_utils diagnostic message

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: add publish

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

technolojin commented:
Request:

In the warn/error message, it would be better to show the condition of diagnostics (ex. processing time 123.4 ms exceeded threshold of 100.0 ms).
So that user can determine whether the system is bad or the threshold is not good.

vividf committed:
chore: add detail message

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf commented:
@technolojin 
Thanks for your suggestion!
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10579/commits/b169714e5c00a5623435ddb09c2e1688d1ee836c

mojomex reviewed (changes requested):
The diagnostics are looking good I think. Two minor details:
* could you rename `timestamp` to `cloud_header_timestamp` or similar to clarify that it's not the system time stamp at the time of processing?
* could you add `pipeline_latency` to concat? Intuitively, this should probably be the maximum pipeline latency of all clouds. Or has this been discussed before?

As for the implementation, a lot of the diagnostics have just been added to already super long functions, so I'm afraid this makes this harder to maintain and is not super modular. Given that we also have to do this for the CUDA implementation, what I would like is a few small classes like
* `LatencyDiagnostics` for cloud header timestamp, processing time, pipeline latency
* `PassRateDiagnostics` for input/output count, pass rate
* `DistortionCorrectorDiagnostics` and `CropBoxDiagnostics` for the rest

The idea would be to have only one function call to each of those classes, and then diagnostics can be published. This would make it very easy to keep pointcloud_preprocessor and cuda_pointcloud_preprocessor in sync.

vividf committed:
chore: const for time difference

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
refactor: structure diagnostics to class

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: const reference

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: clean logic

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
Merge branch 'main' into feat/add_diag_msg

vividf commented:
Hi @mojomex,
Thank you for the detailed review!
I've addressed all the suggestions you mentioned. Would you mind taking another look when you have a moment?
Thanks again!

mojomex commented:
Hi @vividf, it looks like this needs to be rebased in onder to pass CI :bow:

vividf committed:
Merge branch 'main' into feat/add_diag_msg

manato reviewed (commented):
@vividf 
Thank you for the many improvements in introducing diagnostic functionality. 
I left a small comment on parameter naming. I'd appreciate it if you could consider it

vividf committed:
chore: modify function name

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: update parameter

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf commented on sensing/autoware_pointcloud_preprocessor/schema/crop_box_filter_node.schema.json:
```diff
@@ -40,9 +40,23 @@
           "description": "if true, remove points within the box from the pointcloud; otherwise, remove points outside the box.",
           "default": "false"
+        },
+        "processing_time_threshold": {
```
@manato 
Thanks for your suggestion!
Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10579/commits/14c5492cec44ab6d978548f63cbb30823c25b97f

manato replied:
@vividf 
Thanks for the prompt fixing. LGTM!

mojomex reviewed (commented):
Thanks for your work, I think it's really clean actually. Some very minor naming comments but otherwise great!

One point for discussion:
Do you think it would make sense to move the thresholds (processing time, mismatched points etc.) into the `...Diagnostics` classes themselves and have them call `update_message_and_status` of the interface when calling `add_to_interface`?
The diagnostic interface can aggregate statuses and messages (more severe status wins, messages are concatenated somehow) but I honestly haven't tested it so not sure if it's usable for this.

vividf committed:
chore: move evaluate status into diagnostic

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: fix description for concatenated pointcloud

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf committed:
chore: timestamp mismatch threshold

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

drwnz commented on sensing/autoware_pointcloud_preprocessor/include/autoware/pointcloud_preprocessor/concatenate_data/concatenate_and_time_sync_node.ipp:
```diff
@@ -411,80 +433,75 @@ PointCloudConcatenateDataSynchronizerComponentTemplated<MsgTraits>::find_and_res
-        format_timestamp(advanced_info->timestamp + advanced_info->noise_window));
-    }
+  diagnostics_interface_->add_key_value(
+    "concatenated_cloud_timestamp", format_timestamp(current_concatenate_cloud_timestamp_));
```
```suggestion
    "Concatenated pointcloud timestamp", format_timestamp(current_concatenate_cloud_timestamp_));
```

drwnz reviewed (changes requested):
Since we decided on readable text for key values, I have left some suggestions.
Otherwise, looks good!

vividf commented on sensing/autoware_pointcloud_preprocessor/schema/distortion_corrector_node.schema.json:
```diff
@@ -25,13 +25,25 @@
+          "description": "Threshold in seconds. If the processing time of the node exceeds this value, a diagnostic warning will be issued.",
+          "default": 0.01
+        },
+        "mismatch_fraction_threshold": {
```
addressed in https://github.com/autowarefoundation/autoware_universe/pull/10579/commits/eda8d08aad749c7ec352dbfe76f13d4d9bcac938

vividf commented on sensing/autoware_pointcloud_preprocessor/include/autoware/pointcloud_preprocessor/concatenate_data/concatenate_and_time_sync_node.ipp:
```diff
@@ -411,80 +433,75 @@ PointCloudConcatenateDataSynchronizerComponentTemplated<MsgTraits>::find_and_res
+  if (diagnostic_info.drop_previous_but_late_pointcloud) {
+    level = diagnostic_msgs::msg::DiagnosticStatus::ERROR;
+    message = topic_miss
+      ? "Concatenated pointcloud misses some topics and is not published because it arrived too late"
```
Thanks! Addressed in https://github.com/autowarefoundation/autoware_universe/pull/10579/commits/b892858afa5dd902926f5d12d4217aaf9873eaaa

vividf commented on sensing/autoware_pointcloud_preprocessor/src/distortion_corrector/distortion_corrector.cpp:
```diff
@@ -359,7 +363,7 @@ void DistortionCorrector<T>::undistort_pointcloud(
       twist_stamp = rclcpp::Time(it_twist->header.stamp).seconds();
     }
-    if (std::abs(current_point_stamp - twist_stamp) > 0.1) {
+    if (std::abs(current_point_stamp - twist_stamp) > time_diff) {
```
addressed in https://github.com/autowarefoundation/autoware_universe/commit/eda8d08aad749c7ec352dbfe76f13d4d9bcac938

vividf committed:
chore: fix diagnostic key

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf commented:
@drwnz 
Thanks for the review!
Addressed them in https://github.com/autowarefoundation/autoware_universe/pull/10579/commits/c01ee40f4d305db3b7b8debba28dd7c44d5aad10

vividf committed:
Merge branch 'main' into feat/add_diag_msg

vividf commented:
Updated diagnostics message

![Screenshot from 2025-06-03 09-37-17](https://github.com/user-attachments/assets/883855f5-149e-4dec-a0c5-79df0e658e5f)

![Screenshot from 2025-06-03 09-37-14](https://github.com/user-attachments/assets/0d664931-1729-4777-8c90-01c2acc1e846)


![Screenshot from 2025-06-03 09-37-09](https://github.com/user-attachments/assets/b5ab100c-69cf-40cb-8169-85cd0fb6164f)

![Screenshot from 2025-06-03 09-37-02](https://github.com/user-attachments/assets/fe6cd08b-8394-4d66-b3a2-7e03f9f6ca97)

drwnz reviewed (approved):
LGTM!

mojomex reviewed (approved):
LGTM!

vividf commented:
@YoshiRi @technolojin 
Since some function names are changed in the perception packages, could you review it :) 
Thanks!

technolojin commented:
Did you prepared autoware_launch PR for [voxel_grid_based_euclidean_cluster.param.yaml](https://github.com/autowarefoundation/autoware_launch/blob/13273a0d056f6d486305e29e2210b6af6ee446bb/autoware_launch/config/perception/object_recognition/detection/clustering/voxel_grid_based_euclidean_cluster.param.yaml#L4) update?

vividf commented:
This branch should merge together with 
* https://github.com/autowarefoundation/autoware_launch/pull/1429
* https://github.com/tier4/aip_launcher/pull/461

technolojin reviewed (approved):
LGTM

badai-nguyen reviewed (changes requested):
@vividf 
`ground_segmenation` package is also inherited from `Filter class of filter.cpp` so I think some paramCallback usage also need to be renamed like 
- https://github.com/autowarefoundation/autoware_universe/blob/ece38099fe94d959e98ce582bc04be49c6c1f12b/perception/autoware_ground_segmentation/src/ray_ground_filter/node.cpp#L73
- https://github.com/autowarefoundation/autoware_universe/blob/e875ddaeba0ee4c3b3328f10f717c8244e573b76/perception/autoware_ground_segmentation/src/ransac_ground_filter/node.hpp#L122

vividf committed:
chore: change function naming

Signed-off-by: vividf <yihsiang.fang@tier4.jp>

vividf commented:
@badai-nguyen 
After checking the code again, I actually don't need to change the naming for the perception node because it has no dependency on the filter's param_callback. 

Since I already modified most of the perception node, I think I should also do for this one.
https://github.com/autowarefoundation/autoware_universe/pull/10579/commits/5e09c742663d5c5a919e1097fa92f2e1220574f3

Thanks for your check!

vividf committed:
Merge branch 'main' into feat/add_diag_msg

badai-nguyen reviewed (approved):
> @badai-nguyen After checking the code again, I actually don't need to change the naming for the perception node because it has no dependency on the filter's param_callback.
> 
> Since I already modified most of the perception node, I think I should also do for this one. [5e09c74](https://github.com/autowarefoundation/autoware_universe/commit/5e09c742663d5c5a919e1097fa92f2e1220574f3)
> 
> Thanks for your check!

@vividf Sorry, you are correct. Since the function name are similar, It was my mistake when look at `filterParamCallback` of base class Filter and function `ParamCallback` in derived class `ground_segmentation`
Anyway, thanks for your work for perception part. 
LGTM! :+1:
